Address issues running geopmctl as root

bgeltz opened the pull request:
- Fixes #3352.

cmcantalupo commented on src/Controller.cpp:
```diff
@@ -261,6 +261,10 @@ namespace geopm
         , m_do_restore(false)
     {
+        if (geopm::has_cap_sys_admin()) {
+            throw Exception("Controller cannot be run as root",
```
```suggestion
            throw Exception("Controller cannot be run with CAP_SYS_ADMIN",
```

bgeltz committed:
Remove batch context from MSRFieldConstructor

- Instead add a setter so it can be modified
  after object construction.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Set the batch context after pruning

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Inspect TRL writability during msr-safe allowlist creation

- Requires geopmadmin -a to be run and applied
  twice to get the desired allowlist in place.
- Adds call to system_write_mask() during creation of
  MSRFieldControls.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add verbose error message for platform_get_profile_pids

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Fix MSR tests

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz commented on service/src/MSRFieldControl.cpp:
```diff
@@ -47,9 +51,13 @@ namespace geopm
-
-        m_save_idx = m_msrio->add_read(m_cpu, m_offset, m_save_restore_ctx);
-        m_restore_idx = m_msrio->add_write(m_cpu, m_offset, m_save_restore_ctx);
+        uint64_t offset_mask = m_msrio->system_write_mask(m_offset);
```
Add LUT to system_write_mask() to avoid repeatedly calling it after we've done 1 lookup.  The mask won't change.

bgeltz replied:
This already exists in the system_write_mask() implementation.

bgeltz commented on service/src/MSRIOGroup.cpp:
```diff
@@ -1717,6 +1740,19 @@ namespace geopm
+                      << "(): unable to check trl via platform_info: "
+                      << ex.what() << std::endl
+                      << "put this initial allowlist in place and rerun to allow for "
+                      << "platform_info inspection." << std::endl;
```
```suggestion
                      << "PLATFORM_INFO inspection." << std::endl;
```

bgeltz replied:
Change this around so that the TRL check uses the stock MSR driver initially to do the PLATFORM_INFO inspection.
Temporarily modprobe msr, run the check, then rmmod msr, even if msr-safe is present.

bgeltz commented on src/Controller.cpp:
```diff
@@ -261,6 +261,10 @@ namespace geopm
+        if (geopm::has_cap_sys_admin()) {
+            throw Exception("Controller cannot be run with CAP_SYS_ADMIN.  Try again as a "
+                            "non-root user.", GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
```
Remove as running the Controller as root works now.

bgeltz committed:
Make msr_allowlist testable

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add allowlist C API

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Remove allowlist and cpuid features of geopmadmin

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Add docu

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Fix geopmaccess tests

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz commented on service/geopmdpy/access.py:
```diff
@@ -301,8 +309,26 @@ def read_names(self, fid):
+
+        """
+        global _dl
+        name_max = 4096
```
```suggestion
        data_max = 2MB
```

bgeltz commented on service/geopmdpy/access.py:
```diff
@@ -388,11 +417,14 @@ def run(self, is_write, is_all, is_control, group, is_default, is_delete,
                     output = self.get_user_controls()
                 else:
                     output = self.get_user_signals()
+
```
```suggestion
```

bgeltz commented on service/docs/source/geopmaccess.1.rst:
```diff
@@ -349,6 +347,50 @@ and combine access lists when writing to a shared mount point.
+
+.. code-block:: bash
+
+    # Run the following with root or sudo:
```
If msr is already loaded, avoid modprobe msr and rmmod.  Check if there's an RC with modprobe that can be examined.

bgeltz replied:
modprobe has a 0 RC even if the module is already loaded 😮

bgeltz committed:
Address review feedback

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

bgeltz committed:
Update sphinx year to 2024

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
